gateway/sqlite.go

```diff
@@ -0,0 +1,167 @@
+// Package gateway implements the sx1302 gateway module.
+package gateway
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	// github.com/mattn/go-sqlite3 is for sqlite.
+	_ "github.com/mattn/go-sqlite3"
+)
+
+var (
+	errTXTMigration = errors.New("error migrating device backup to sqlite. please reset your sensors")
+	errNoDeviceInDB = errors.New("error device not found")
+	errNoDB         = errors.New("error device file not found")
+)
+
+// Create or open a sqlite db file used to save device data across restarts.
+func (g *gateway) setupSqlite(ctx context.Context) error {
+	moduleDataDir := os.Getenv("VIAM_MODULE_DATA")
+
+	filePathDB := filepath.Join(moduleDataDir, "devicedata.db")
+	db, err := sql.Open("sqlite3", filePathDB)
+	if err != nil {
+		return err
+	}
+	// create the table if it does not exist
+	cmd := `create table if not exists `
+	keys := `
+	devices(devEui TEXT NOT NULL PRIMARY KEY, appSKey TEXT, nwkSKey TEXT, devAddr TEXT, fCntDown INTEGER, nodeName TEXT);
+	`
+	sqlStmt := cmd + keys
+	if _, err = db.ExecContext(ctx, sqlStmt); err != nil {
+		return err
+	}
+	g.db = db
+
+	// check if the machine has an old devicedata file for us to migrate
+	filePathTXT := filepath.Join(moduleDataDir, "devicedata.txt")
+	if _, err := os.Stat(filePathTXT); err == nil {
+		txtDevices, err := readFromFile(filePathTXT)
+		if err != nil {
+			return errTXTMigration
+		}
+		// move devices found from the old backup logic into the sqlite db
+		for _, device := range txtDevices {
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+			err = g.insertOrUpdateDeviceInDB(ctx, device)
+			if err != nil {
+				return errTXTMigration
+			}
+		}
+		if err := os.Remove(filePathTXT); err != nil {
+			return errTXTMigration
+		}
+	}
+
+	return nil
+}
+
+func (g *gateway) insertOrUpdateDeviceInDB(ctx context.Context, device deviceInfo) error {
+	if g.db == nil {
+		return errNoDB
+	}
+	cmd := `insert or replace into `
+	keys := `devices (devEui, appSKey, nwkSKey, devAddr, fCntDown, nodeName) VALUES(?, ?, ?, ?, ?, ?);`
+	_, err := g.db.ExecContext(ctx, cmd+keys,
+		device.DevEUI,
+		device.AppSKey,
+		device.NwkSKey,
+		device.DevAddr,
+		device.FCntDown,
+		device.NodeName,
+	)
+	return err
+}
+
+func (g *gateway) findDeviceInDB(ctx context.Context, devEui string) (*deviceInfo, error) {
+	if g.db == nil {
+		return nil, errNoDB
+	}
+	devEui = strings.ToUpper(devEui)
+	newDevice := deviceInfo{}
+	if err := g.db.QueryRowContext(ctx, "select * from devices where devEui = ?",
+		devEui).Scan(&newDevice.DevEUI, &newDevice.AppSKey, &newDevice.NwkSKey,
+		&newDevice.DevAddr, &newDevice.FCntDown, &newDevice.NodeName); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return &deviceInfo{}, errNoDeviceInDB
+		}
+		return &deviceInfo{}, err
+	}
+	return &newDevice, nil
+}
+
```

Comment:
I recommend not returning pointers if you use plane old structs elsewhere.
```suggestion
func (g *gateway) findDeviceInDB(ctx context.Context, devEui string) (deviceInfo, error) {
	if g.db == nil {
		return nil, errNoDB
	}
	var zero deviceInfo
	devEui = strings.ToUpper(devEui)
	newDevice := deviceInfo{}
	if err := g.db.QueryRowContext(ctx, "select * from devices where devEui = ?",
		devEui).Scan(&newDevice.DevEUI, &newDevice.AppSKey, &newDevice.NwkSKey,
		&newDevice.DevAddr, &newDevice.FCntDown, &newDevice.NodeName); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return zero, errNoDeviceInDB
		}
		return zero, err
	}
	return newDevice, nil
}

```